### PR TITLE
Track-caller panics

### LINKED ISSUE #972: Add `#[track_caller]` judiciously
Add `#[track_caller]` to some panicking methods. Slicing methods seem the most prudent (Index/Mut does not apply, sorry).

Needs benchmarks to investigate impact, preferably holistic benchmarks.

### PR DESCRIPTION
**Pull request by xd009642**
Add track caller to any function, method, trait that has a documented panics section. I also used cargo fmt to tidy up my semi-auto track_caller labelling so there may be other changes but they'll just be formatting.

Still need to assess the performance impact of all these track callers!

Fixes #972

### CODE REVIEW

**Comment by xd009642** [comment]
So I tried the benchmarks in the repo just to see if there were any glaring issues but the results seemed very noisy and an even mix of faster than master or slower than master. I'll create some more directed benchmarks and try to figure out what the impact is

File: src/arraytraits.rs [hunk 1]
```diff
@@ -47,6 +47,7 @@ where
 {
     type Output = S::Elem;
     #[inline]
+    #[track_caller]
     fn index(&self, index: I) -> &S::Elem {
```
Comment (bluss) [inline src/arraytraits.rs:51]: It doesn't work on trait methods I thought?

File: ndarray-rand/src/lib.rs [hunk 2]
```diff
@@ -35,7 +35,7 @@ use crate::rand::seq::index;
 use crate::rand::{thread_rng, Rng, SeedableRng};
 
 use ndarray::{Array, Axis, RemoveAxis, ShapeBuilder};
-use ndarray::{ArrayBase, DataOwned, RawData, Data, Dimension};
+use ndarray::{ArrayBase, Data, DataOwned, Dimension, RawData};
```
Comment (bluss) [inline ndarray-rand/src/lib.rs:38]: Formatting changes for the whole crate should be in a different PR, see #922

Comment (xd009642) [reply ndarray-rand/src/lib.rs:38]: Yeah I tried not to commit anything that I hadn't changed but I'll go through and revert what I can. It's just a bit faffy when cargo fmt is basically part of my workflow :sweat_smile:

File: src/arraytraits.rs (hunk 1 above)
Comment (xd009642) [reply src/arraytraits.rs:51]: https://doc.rust-lang.org/reference/attributes/codegen.html Looks like it does, but I should just put it on the trait definition not implementations

**Comment by bluss** [comment]
If we can find, a non-microbenchmark like mentioned #972

File: src/arraytraits.rs (hunk 1 above)
Comment (xd009642) [reply src/arraytraits.rs:51]: Ah but Index and IndexMut are in the standard library... And already tracked there! I'll remove :+1:

File: src/arraytraits.rs [hunk 3]
```diff
@@ -182,6 +183,7 @@ where
     /// let array = Array::from_iter((0..5).map(|x| x * x));
     /// assert!(array == arr1(&[0, 1, 4, 9, 16]))
     /// ```
+    #[track_caller]
```
Comment (bluss) [inline src/arraytraits.rs:186]: Should not have track caller - it's a trait that doesn't support it, and the panic reason is very obscure, not worth it.

**Comment by xd009642** [comment]
Okay so I created this microbenchmark of slicing:

```rust
use criterion::{black_box, criterion_group, criterion_main, Criterion};
use ndarray::{s, Array3};
use ndarray_rand::RandomExt;
use ndarray_rand::rand_distr::Uniform;

fn slice_benchmark(c: &mut Criterion) {
    let data: Array3<f64> = Array3::random((64,64,64), Uniform::new(0., 10.));

    c.bench_function("slice", |b| b.iter(|| {
        data.slice(black_box(s![.., 24..50, ..])).shape();
    }));
}

criterion_group!(benches, slice_benchmark);
criterion_main!(benches);
```

Results:

**0.15.0**
slice                   
time:   [34.281 ns 34.723 ns 35.189 ns]                   
Found 9 outliers among 100 measurements (9.00%)
  4 (4.00%) high mild
  5 (5.00%) high severe

**This PR**
slice                   
time:   [32.815 ns 33.108 ns 33.458 ns]                   
change: [-8.6198% -5.7554% -3.0651%] (p = 0.00 < 0.05)
Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe

Results folder here:

[criterion_results.zip](https://github.com/rust-ndarray/ndarray/files/6273524/criterion_results.zip)

**Comment by xd009642** [comment]
Personally, I reckon the shift in the results between the two is probably down to something else scheduled by my OS taking some cycles away and these two versions can be considered equivalent in performance (just from this benchmark).

Whether this microbenchmark is good enough to check the impact of track caller is down to someone more familiar with ndarray internals :slightly_smiling_face:

**Comment by xd009642** [comment]
Any thoughts about the benchmark @bluss ?

**Comment by bluss** [comment]
The maintainers have asked for judicious use of the track caller attribute and applying it everywhere is funny to me :slightly_smiling_face:  because it's basically the literal opposite by the dictionary of what judicious means.

Maintainers have asked for a holistic benchmark - maybe a real program that uses ndarray for a lot of varied operations.

Just some real talk :slightly_smiling_face:. If the contributor doesn't want to be *judicious* or whatever, then the maintainer instead has to judge line by line if the additions make sense. And I don't have the investment to do that at the moment, sorry. In short this PR is stalling a bit since it seems to go a bit against every instruction that I have given to try to guide its resolution.

**Comment by bluss** [comment]
A general nice github feature: please use a line like "Fixes #972" to the PR description when making a change that closes an issue. Issues should be closed automatically on merge, that's the nice thing.

**Comment by xd009642** [comment]
So I didn't apply it everywhere the PR title is a bit inaccurate in that sense. I only applied it on public API functions figuring only panics that can be caused directly by user code should be tracked. But I'll go through them and aim to reduce the scope further.

Oh I misread the non-microbenchmark as microbenchmark sorry that ones on me. I was looking at some projects like linfa to see if I could find bigger example projects that used some methods that panicked but I didn't find anything that looked useful. I'll go back to that.

**Comment by bluss** [comment]
Updated to master and conflicts resolved, I hope.

**Comment by bluss** [comment]
Squashed to clean up history. Current merge queue config will create a merge commit at the end too.

**Comment by bluss** [comment]
Thanks, sorry for the delay.
